graph/src/components/store.rs

```diff
@@ -194,19 +211,60 @@ impl EntityChange {
 /// any specific blocks, `StoreEvents` for it are generated as soon as they are
 /// written to the store.
 pub struct StoreEvent {
-    pub source: EventSource,
-    pub changes: Vec<EntityChange>,
+    // The tag is only there to make it easier to track StoreEvents in the
+    // logs as they flow through the system
+    pub tag: usize,
+    pub changes: HashSet<EntityChange>,
 }
 
-/// A boxed `StoreEventStream`
-pub type StoreEventStreamBox = Box<Stream<Item = StoreEvent, Error = ()> + Send>;
+impl StoreEvent {
+    pub fn new(changes: Vec<EntityChange>) -> StoreEvent {
+        static NEXT_TAG: AtomicUsize = AtomicUsize::new(0);
+
+        let tag = NEXT_TAG.fetch_add(1, Ordering::Relaxed);
+        let changes = changes.into_iter().collect();
+        StoreEvent { tag, changes }
+    }
+
+    /// Extend `ev1` with `ev2`. If `ev1` is `None`, just set it to `ev2`
+    fn accumulate(logger: &Logger, ev1: &mut Option<StoreEvent>, ev2: StoreEvent) {
+        if let Some(e) = ev1 {
+            trace!(logger, "Adding changes to event";
+                           "from" => ev2.tag, "to" => e.tag);
+            e.changes.extend(ev2.changes);
+        } else {
+            ev1.replace(ev2);
```

Comment:
This is just `ev1 = Some(ev2)`.

Reply:
I can't find a way to do that: just `ev1 = Some(ev2)` makes rustc complain because `ev1` is `&mut`, and `ev1 = &mut Some(ev2)` makes it complain that it can't assign twice to `ev1` and that the reference doesn't live long enough. This really needs to change the contents of `ev1` in the caller as that's what we are accumulating into. And returning the combination of `ev1` and `ev2` to the caller isn't desirable since `changes` can become a long list that we'd need to copy every time we combine events.

Reply:
Fixed it with `*ev1 = Some(ev2)` as you suggested in Discord.